subprojects/core/src/main/java/org/gradle/api/internal/GradleInternal.java

```diff
@@ -151,6 +159,8 @@ public interface GradleInternal extends Gradle, PluginAwareInternal {
      */
     ClassLoaderScope baseProjectClassLoaderScope();
 
+    void baseProjectClassLoaderLocked(Action<? super GradleInternal> action);
```

Comment:
💅 Can we add a javadoc here? It'd be nice to clarify it's about adding a callback. It'd also be nice to specify the contract as to when these callbacks are invoke relative to other callbacks, such as `projectsLoaded`, etc

I think at this point, we should explicitly clarify the thread-safety of callback adding methods. The current implementation of this does not seem to be thread-safe. It might not need to be now, but we've seen how this can bite us with Isolated Projects.

I personally find this naming convention non-intuitive -- the fact that it will **add** an action to be executed later. But we use it for many other callbacks, unfortunately.

I think a javadoc would be a good middleground